pkg/reconciler/example/route.go

```diff
@@ -0,0 +1,419 @@
+}
+
+func (s *DesiredRoute) PrimaryKey() []byte {
+	return s.RouteID().Key()
+}
+
+func (s *DesiredRoute) RouteID() tables.RouteID {
+	return tables.RouteID{
+		Table:     s.Route.Table,
+		LinkIndex: s.Route.LinkIndex,
+		Dst:       s.Route.Dst,
+	}
+}
+
+func (s *DesiredRoute) GetStatus() reconciler.Status {
+	return s.Status
+}
+
+func (s *DesiredRoute) WithStatus(newStatus reconciler.Status) *DesiredRoute {
+	s2 := *s
+	s2.Status = newStatus
+	return &s2
+}
+
+func (d *DesiredRoute) toNetlinkRoute(linkIndex int) *netlink.Route {
+	return &netlink.Route{
+		Table:     d.Route.Table,
+		LinkIndex: linkIndex,
+		Dst:       prefixToIPNet(d.Route.Dst),
+		Src:       d.Route.Src.AsSlice(),
+		Gw:        d.Route.Gw.AsSlice(),
+		Scope:     netlink.Scope(d.Route.Scope),
+		Protocol:  linux_defaults.RTProto,
+	}
+}
+
+var (
+	RouteIDIndex = statedb.Index[*DesiredRoute, tables.RouteID]{
+		Name: "id",
+		FromObject: func(s *DesiredRoute) index.KeySet {
+			return index.NewKeySet(s.PrimaryKey())
+		},
+		FromKey: tables.RouteID.Key,
+		Unique:  true,
+	}
+	RouteOwnerIndex = statedb.Index[*DesiredRoute, string]{
+		Name: "owner",
+		FromObject: func(s *DesiredRoute) index.KeySet {
+			return index.NewKeySet(index.String(s.Owner))
+		},
+		FromKey: index.String,
+		Unique:  false,
+	}
+	RouteStatusIndex = reconciler.NewStatusIndex[*DesiredRoute]()
+)
+
+//
+// Route reconciliation target
+//
+
+type routeTestTarget struct {
+	netlinkHandle *netlink.Handle
+	devices       statedb.Table[*tables.Device]
+	routes        statedb.Table[*tables.Route]
+}
+
+func (routeTestTarget) Init(context.Context) error {
+	return nil
+}
+
+func (t *routeTestTarget) Delete(_ context.Context, txn statedb.ReadTxn, desired *DesiredRoute) error {
+	fmt.Printf("Delete: %v\n", desired)
+
+	_, _, ok := t.routes.First(txn, tables.RouteIDIndex.Query(desired.RouteID()))
+	if !ok {
+		// Route already gone.
+		return nil
+	}
+
+	linkIndex := desired.Route.LinkIndex
+
+	// TODO
+	if desired.OptDeviceName != "" {
+		dev, _, ok := t.devices.First(txn, tables.DeviceNameIndex.Query(desired.OptDeviceName))
+		if !ok {
+			return fmt.Errorf("device %q not found for route to %q", desired.OptDeviceName, desired.Route.Dst.String())
+		}
+		linkIndex = dev.Index
+	}
+
+	if err := t.netlinkHandle.RouteDel(desired.toNetlinkRoute(linkIndex)); err != nil {
+		return fmt.Errorf("failed to delete route to %q owned by %q: %w",
+			desired.Route.Dst.String(),
+			desired.Owner,
+			err)
+	}
+	return nil
+}
+
+// Sync implements reconciler.Target
+func (t *routeTestTarget) Sync(ctx context.Context, txn statedb.ReadTxn, iter statedb.Iterator[*DesiredRoute]) (outOfSync bool, err error) {
+	// TODO: Might want to trigger sync when devices change their state.
+	// (e.g. go UP from DOWN).
+	fmt.Printf("Sync\n")
+
+	var errs []error
+
+	for desired, _, ok := iter.Next(); ok; desired, _, ok = iter.Next() {
+		actual, _, ok := t.routes.First(txn, tables.RouteIDIndex.Query(desired.RouteID()))
```

Comment:
Why "First"?

Reply:
Because we’re only interested in the first match. StateDB supports both unique and non-unique indexes and doesn’t differentiate the query API for these.